File: 2264-largest-3-same-digit-number-in-string/2264-largest-3-same-digit-number-in-string.py
class Solution:
    def largestGoodInteger(self, num: str) -> str:
        '''
            Time complexity = O(n) 
            Space complexity = O(1)
        '''

        n = len(num)
        result = ""

        left, right = 0, 2

        while(right < n) :
            if num[left] == num[left+1] and \
            num[left+1] == num[left+2] and \
            num[left+2] == num[left] :

                if result == "" or num[left] > result[0] :
                    result = num[left] * 3 # string concatenation.
            
            left += 1
            right += 1
        
        return result

File: 2264-largest-3-same-digit-number-in-string/2264-largest-3-same-digit-number-in-string.java
class Solution {
    public String largestGoodInteger(String num) {
        /*
            Two Pointer Approach :-
                Time complexity = O(n), 
                Space complexity = O(1) 
        */
        
        int n = num.length();
        int left = 0, right = 2;
        String result = "";

        while(right < n){
            char numCh = num.charAt(left);
            char numChNext = num.charAt(left + 1);
            char numChNextNext = num.charAt(left + 2);

            if(numCh == numChNext && numChNext == numChNextNext && numChNextNext == numCh)
            {
                if(result.isEmpty() || numCh > result.charAt(0))
                {
                    result = "" + numCh + numCh + numCh;
                }
            }

            left++;
            right++;
        }

        return result;
    }
}